From 3f7a1aa7a4f3b41530703306dad57640607929ad Mon Sep 17 00:00:00 2001
From: Lionel Ulmer <lionel.ulmer@intel.com>
Date: Fri, 12 Jun 2015 16:41:09 +0200
Subject: [PATCH] usb: gadget: release spinlock before calling into TTY code

Function gs_start_tx calls tty_wakeup with the port_lock
spinlock locked. This lead to spinlock recursion as the TTY code
can - in some rare cases - call back into the USB serial code.

Fix the issue by releasing the spinlock before calling into the
TTY code.

Change-Id: I931305150d51bebf6b1e6b5e0ae9a69f668e6cda
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-39313
Signed-off-by: Lionel Ulmer <lionel.ulmer@intel.com>
---
 drivers/usb/gadget/u_serial.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/gadget/u_serial.c b/drivers/usb/gadget/u_serial.c
index b131751933cb..21dff1c71933 100644
--- a/drivers/usb/gadget/u_serial.c
+++ b/drivers/usb/gadget/u_serial.c
@@ -414,8 +414,11 @@ __acquires(&port->port_lock)
 			break;
 	}
 
-	if (do_tty_wake && port->port.tty)
+	if (do_tty_wake && port->port.tty) {
+		spin_unlock(&port->port_lock);
 		tty_wakeup(port->port.tty);
+		spin_lock(&port->port_lock);
+	}
 	return status;
 }
 
-- 
1.9.1

